### ISSUE TITLE
Support NonNullApi/NonNullFields 

### ISSUE THREAD

**Issue by xenoterracide** [uber/NullAway#1084]
Support spring style package annotations https://docs.spring.io/spring-framework/reference/core/null-safety.html gradle-api also has `NonNullApi`, I suspect in lieu of java support this will become more popular. I imagine not supporting it could be problematic when working with these libraries. 

Since fields are rarely exposed I could see `NonNullApi` being treated as a synonym for `Nullmarked` initially being a safe-ish way forward for consumers. You could document it as such ... "this is partial support"... which would serve my purposes as a consumer.

**Comment by msridhar** [comment]
@xenoterracide just confirming, do IntelliJ or other tools already support `@NonNullApi` in the manner that you're proposing here?

**Comment by xenoterracide** [comment]
I'm not certain what you mean by "the manner" and I admit I haven't done comprehensive analysis. It does at least have some support, although I think I've found a bug based on this analysis.... the [2 arg findField](https://github.com/spring-projects/spring-framework/blob/main/spring-core/src/main/java/org/springframework/util/ReflectionUtils.java#L580-L610) I see, both parameters should not be nullable... I think that's an intellij bug that I don't care enough about to report.

![Screenshot from 2024-12-12 13-40-15](https://github.com/user-attachments/assets/dc32783c-8624-43e6-bec7-024295db4a9d)

you can see the package-info is annotated with these annotations. If both are supplied they should be equivalent to putting nullmarked on the package. https://github.com/spring-projects/spring-framework/blob/main/spring-core/src/main/java/org/springframework/util/package-info.java

I care about this detection only for 3rd party packages, but I don't know if you have any way to tell the difference between compiling against a jar vs the code you're compiling. Point being more that I don't think if you're only supporting this for detection on 3rd party libraries I'm not certain you have to worry about the differences between having and not having `NonNullFields` when using `NonNullApi`.

**Comment by msridhar** [comment]
Right I just meant other tools are treating `@NonNullApi` and the like as kind of equivalent to `@NullMarked`.  Thanks!

**Comment by xenoterracide** [comment]
More or less... As stated there's an interaction with NonNullFields as it's not supposed to imply that. If you only had non-null API then the fields would be considered nullable by default and you can imagine the vice versa as well...

**Comment by xenoterracide** [comment]
More or less... As stated there's an interaction with NonNullFields as it's not supposed to imply that. If you only had non-null API then the fields would be considered nullable by default and you can imagine the vice versa as well...

**Comment by xenoterracide** [comment]
p.s. I have created tickets for libraries that I'm aware of using `NonNullApi` to add `Nullmarked`

**Comment by msridhar** [comment]
@xenoterracide I did a quick hack to add this support in #1095.  Could you pull the branch there, install a snapshot (by running `ORG_GRADLE_PROJECT_RELEASE_SIGNING_ENABLED=false ./gradlew publishToMavenLocal`), and see if it works as you expect?

Before landing this change I would need to do some further testing including on Uber code, as this has the potential to make NullAway report a bunch of new warnings.  We may have to put it behind a flag at first.

**Comment by sdeleuze** [comment]
Spring Framework 5.x and 6.x null-safety annotations including `@NonNullApi` / `@NonNullFields` are already supported by NullAway due to their JSR 305 meta annotations. Spring Framework 7.x will migrate its whole codebase to JSpecify annotations.

Also `@NonNullApi`/`@NonNullField` and `@NullMarked` have subtle but very important differences, as the former apply to parameter parameters/return values/fields while the latter applies to the use of types. The semantics of nullability for arrays, varags and generic types are also totally different.

As consequence, I am strongly recommending declining this feature request.

**Comment by msridhar** [comment]
Thanks a lot @sdeleuze!  I will then close this in favor of #633.  @xenoterracide we can discuss further on #633, but given our limited resources, I am leaning towards not prioritizing support for `@TypeQualifierNickname` right now.

**Comment by xenoterracide** [comment]
I mean I don't have much to say other than that's great for spring internally but that's not great for projects that use spring. I don't know if Gradle has any plans to do the same thing though. Also last I checked the spring migration to Jspecify was being held up by the fact that jspecify 1.0 decided to remove the meta annotations. Sometimes conversations don't trickle into tickets though; also I could be forgetting that something was changed in that ticket.
